build/configs/common_download.sh

```diff
@@ -0,0 +1,390 @@
+
+flash_ota=false
+##Utility function to get partition index ##
+function get_partition_index()
+{
+	for idx in ${!parts[@]}; do
+		if [[ ${parts[$idx],,} == ${1,,} ]]; then
+			if [[ $flash_ota == true ]]; then
+				## In this case, try to find 2nd kernel partition for OTA
+				flash_ota=false
+				continue;
+			fi
+			echo $idx
+			exit 1
+		fi
+	done
+	echo -1
+}
+
+##Help utility##
+function dwld_help()
+{
+        cat <<EOF
+	HELP:
+		make download ALL or [PARTITION(S)]
+		make download ERASE ALL or erase [PARTITION(S)]
+
+	PARTITION(S):
+		[${uniq_parts[@]}]  NOTE:case sensitive
+
+	For examples:
+		make download ALL
+		make download ERASE ALL
+		make download kernel
+		make download erase kernel
+		make download ota
+		make download erase ota
+		make download erase ss
+		make download smartfs
+		make download erase userfs
+EOF
+}
+
+##Utility function to read partitions from .config or Kconfig##
+function get_configured_partitions()
+{
+	local configured_parts
+	# Read Partitions
+	if [[ -z ${CONFIG_FLASH_PART_NAME} ]];then
+
+		configured_parts=`grep -A 2 'config FLASH_PART_NAME' ${PARTITION_KCONFIG} | sed -n 's/\tdefault "\(.*\)".*/\1/p'`
+	else
+		configured_parts=${CONFIG_FLASH_PART_NAME}
+	fi
+
+	echo $configured_parts
+}
+
+##Utility function to get the partition sizes from .config or Kconfig
+function get_partition_sizes()
+{
+	local sizes_str
+	#Read Partition Sizes
+	if [[ -z ${CONFIG_FLASH_PART_SIZE} ]]
+	then
+		sizes_str=`grep -A 2 'config FLASH_PART_SIZE' ${PARTITION_KCONFIG} | sed -n 's/\tdefault "\(.*\)".*/\1/p'`
+	else
+		sizes_str=${CONFIG_FLASH_PART_SIZE}
+	fi
+
+	echo $sizes_str
+}
+
+download_specific_partition()
+{
+	TARGET=$1
+	if [[ ${TARGET} == "ota" || ${TARGET} == "OTA" ]];then
+		TARGET="kernel"
+		flash_ota=true
+	fi
+	partidx=$(get_partition_index ${TARGET})
+	if [[ "${partidx}" < 0 ]];then
+		echo "Not supported"
+		dwld_help
+		exit 1
+	fi
+
+	# Get a filename and Download a file
+	echo ""
+	echo "============================="
+	if [[ $1 == "ota" || $1 == "OTA" ]];then
+		echo "Downloading Kernel OTA binary"
+	else
+		echo "Downloading ${parts[$partidx]} binary"
+	fi
+	echo "============================="
+	exe_name=$(get_executable_name ${parts[$partidx]})
+	board_download $TTYDEV ${offsets[$partidx]} ${exe_name} ${sizes[partidx]}
+
+	echo ""
+	echo "Download $exe_name COMPLETE!"
+}
+
+download_all()
+{
+	echo "Starting Download..."
+	found_kernel=false
+
+	for partidx in ${!parts[@]}; do
+
+		if [[ "${CONFIG_APP_BINARY_SEPARATION}" != "y" ]];then
+			if [[ "${parts[$partidx]}" == "userfs" ]];then
+				continue
+			fi
+		fi
+
+		if [[ "${parts[$partidx]}" == "kernel" ]];then
+			if [[ $found_kernel == true ]];then
+				continue
+			fi
+			found_kernel=true
+		fi
+
+		if [[ "${parts[$partidx]}" == "ftl" ]];then
+			continue
+		fi
+
+		if [[ "${parts[$partidx]}" == "config" ]];then
+			continue
+		fi
+
+		exe_name=$(get_executable_name ${parts[$partidx]})
+		[ "No Binary Match" = "${exe_name}" ] && continue
```

Comment:
If we have a partition and no corresponding binary, then we better print a warning message.